internal/disk/path_trie.go

```diff
@@ -0,0 +1,113 @@
+package disk
+
+import (
+	"sort"
+	"strings"
+)
+
+func pathTrieSplitPath(path string) []string {
+	path = strings.Trim(path, "/")
+	if path == "" {
+		return []string{}
+	}
+
+	return strings.Split(path, "/")
+}
+
+type PathTrie struct {
+	Name    []string
+	Paths   []*PathTrie
+	Payload interface{}
+}
+
+func (trie *PathTrie) match(path []string) int {
+	if len(trie.Name) > len(path) {
+		return 0
+	}
+
+	for i := range trie.Name {
+		if path[i] != trie.Name[i] {
+			return i
+		}
+	}
+
+	return len(trie.Name)
+}
+
+func (root *PathTrie) Lookup(path string) (*PathTrie, []string) {
+
+	if len(root.Name) != 0 {
+		panic("programming error: lookup on non-root trie node")
+	}
+
+	elements := pathTrieSplitPath(path)
+
+	if len(elements) == 0 {
+		return root, elements
+	}
+
+	return root.get(elements)
+}
+
+func (trie *PathTrie) get(path []string) (*PathTrie, []string) {
+	if len(path) < 1 {
+		panic("programming error: expected root node")
+	}
+
+	var node *PathTrie
+	var prefix int
+	for i := range trie.Paths {
+		if prefix = trie.Paths[i].match(path); prefix != 0 {
```

Comment:
If the `trie.Paths[i]` contains more than one directory name in `Name`, then it is not enough to just check that the returned `prefix` is non-zero. If the `trie.Paths[i]` `Name` contains e.g. `[]string{"lib", "osbuild"}` and the `path` is `[]string{"lib", "chrony"}`, they you'll get non-zero `prefix`, but I would say that it is not a match. The `prefix` should not be really lower than the node's `len(Name)` to call it a match.